state-machine.md

```diff
@@ -51,6 +45,44 @@ type VMContext interface {
 
 If the execution completes successfully, changes to the state tree are saved. Otherwise, the message is marked as failed, and any state changes are reverted.
 
+```go
+func ApplyMessage(st StateTree, msg Message) MessageReceipt {
+    st.Snapshot()
+    fromActor := st.GetActor(msg.From)
+    
+    totalCost := msg.Value + (msg.GasLimit * msg.GasPrice)
+    if fromActor.Balance < totalCost {
+        Fatal("not enough funds")
+    }
+    
+    st.DeductFunds(msg.From, totalCost)
```

Comment:
Deducting maximum gas ahead of time and then refunding what's not spent is significantly more complicated to implement correctly than simply deducting spent gas after execution. Since a message execution cannot trigger an account actor to spend more funds, the outcome will always be the same. I've never understood the justification for this.

Reply:
I don't see how this is significantly more complicated to do. It actually seems about the same complexity, with some nice future proofing guarantees that make it easier to reason about the safety of execution.

Reply:
Ok. Slightly more complicated (charge limit then either revert and re-charge or refund is more complicated than simply charging at the end). What I don't get is the future proofing guarantees, and why it's easier to reason about safety?

Reply:
For one, it allows other actors you interact with in your calls to reason about account balances.

re: safety, If you withdraw the funds up-front, then you can be sure the actor will be able to pay for its execution no matter what. In the event that its possible for actors to internally choose the amounts that they send to others (its not right now) then withdrawing up front strictly prevents those actors from spending money that would otherwise be used to pay for execution.